cli/docker/container_manager.go

```diff
@@ -0,0 +1,545 @@
+	released bool
+}
+
+type DockerLogger struct {
+	logger  log.Logger
+	secrets []string
+}
+
+func (dl *DockerLogger) Infof(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Info(redacted)
+}
+
+func (dl *DockerLogger) Errorf(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Error(redacted)
+}
+
+func (dl *DockerLogger) Warnf(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Warn(redacted)
+}
+
+func (dl *DockerLogger) Redact(s string) (string, error) {
+	src := bytes.NewReader([]byte(s))
+	dstBuf := new(bytes.Buffer)
+	logger := log.NewUtilsLogAdapter()
+	redactWriterDst := redactwriter.New(dl.secrets, dstBuf, &logger)
+
+	if _, err := io.Copy(redactWriterDst, src); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, stream copy failed: %s", err)
+	}
+	if err := redactWriterDst.Close(); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, closing the stream failed: %s", err)
+	}
+
+	redactedValue := dstBuf.String()
+	return redactedValue, nil
+}
+
+func NewContainerManager(logger log.Logger, secrets []string) *ContainerManager {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logger.Warnf("Docker client failed to initialize (possibly running on unsupported stack): %s", err)
+	}
+
+	return &ContainerManager{
+		logger: DockerLogger{
+			logger:  logger,
+			secrets: secrets,
+		},
+		workflowContainers: make(map[string]*RunningContainer),
+		serviceContainers:  make(map[string][]*RunningContainer),
+		client:             dockerClient,
+	}
+}
+
+func (cm *ContainerManager) Login(container models.Container, envs map[string]string) error {
+	if container.Credentials.Username != "" && container.Credentials.Password != "" {
+		cm.logger.Infof("ℹ️ Logging into docker registry: %s", container.Image)
+		cm.logger.Infof("ℹ️ Logging into docker registry: %s", container.Image)
+
+		resolvedPassword := resolveEnvVariable(container.Credentials.Password, envs)
+		args := []string{"login", "--username", container.Credentials.Username, "--password", resolvedPassword}
+
+		if container.Credentials.Server != "" {
+			args = append(args, container.Credentials.Server)
+		} else {
+			args = append(args, container.Image)
+		}
+
+		cm.logger.Infof("ℹ️ Running command: docker %s", strings.Join(args, " "))
+
+		out, err := command.New("docker", args...).RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			cm.logger.Errorf(out)
+			return fmt.Errorf("run docker login: %w", err)
+		}
+	}
+	return nil
+}
+
+func (cm *ContainerManager) StartWorkflowContainer(
+	container models.Container,
+	workflowID string,
+	envs map[string]string,
+) (*RunningContainer, error) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	containerName := fmt.Sprintf("workflow-%s", workflowID)
```

Comment:
I suggest adding the `bitrise-` prefix too. In local and self-hosted envs there could be other containers and uses of Docker.